### ISSUE TITLE
Selecting the best trial based on a different criterium?

### ISSUE THREAD

**Issue by konradsemsch** [optuna/optuna#2273]
Hello Optuna Team!

Is it possible to select the best trial also based on a different criterium? As far as I understand by looking at most examples, the best trial is selected based on the mean of results and the max/min values it always chosen. However, I've also seen other ways of selecting the best trial (e.g. [tune in R](https://tune.tidymodels.org/reference/show_best.html)), where simpler models are chosen by sacrificing a bit of performance. This may result in a better (in the meaning of less volatile) estimator. Some examples might be:

- "one-standard error rule" (Breiman _el at, 1984) that selects the most simple model that is within one standard error of the numerically optimal results
- selects the most simple model whose loss of performance is within some acceptable limit.

Looking forward to your view on this!

**Comment by blaney83** [comment]
Hey! Not the Optuna Team, but I thought I would make a suggestion. If you are looking to optimize something other than your standard model performance (e.g. loss or other), have you considered any of the following:

- changing the value returned from each Trial to be a weighted composite value which balances both the model measure (e.g. loss) and your other desired goal (e.g. measure of simplicity)
- optimizing for multiple objectives. Optuna has a multi-objective optimization framework which allows you to specify multiple criteria which should be maximized/minimized.
- optimizing for your main objective (again, loss or other), but saving your secondary metric as a `user_attr` on the trial. Following the completion of your study, you could then select the top 'X' performers and observer the value of their secondary metric, choosing a more stable configuration. (Using the study => dataframe API should make sorting through trials easier here).

Not knowing the "official" answer, these are some of the approaches I would consider- each with it's own trade offs. I'm interested to know more and if any of these covers your usecase.

**Comment by hvy** [comment]
I could perhaps add that we have several samplers that support constrained optimization as well. You could for instance use your model performance as the main objective and define a slack constraint (threshold) on the model complexity. This is an example using the `BoTorchSampler` https://github.com/optuna/optuna/blob/master/examples/multi_objective/botorch_simple.py If you are able to use the latest `master` (or v2.5.0 when it is released), you could also try this with the [`NSGAIISampler`](https://optuna.readthedocs.io/en/latest/reference/generated/optuna.samplers.NSGAIISampler.html).

**Comment by konradsemsch** [comment]
Hi @blaney83 @hvy - thank you for your answers and suggestions! Let me give you some more context maybe:

- we have a `LightGbm` classification model
- at this stage we're not optimising the parameters of the model itself (using default values), but certain properties of the data pipeline through cross-validation with `sklearn`

Here's a high-level sneak peak:

```python

def _study_preprocess_search(self, trial):
        # Define parameters
        self._parameters = {
            "step_model__objective": "binary",
            "step_model__metric": "auc",
            "step_model__verbosity": -1,
            "step_model__boosting_type": "gbdt",

            "step_preprocess_data__branch_feat_date__step_date_format__year": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_date__step_date_format__year",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_date__step_date_format__quarter": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_date__step_date_format__quarter",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_date__step_date_format__month": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_date__step_date_format__month",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_date__step_date_format__week_of_month": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_date__step_date_format__week_of_month",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_date__step_date_format__week_of_year": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_date__step_date_format__week_of_year",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_date__step_date_format__day_of_week": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_date__step_date_format__day_of_week",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_date__step_date_format__day_of_month": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_date__step_date_format__day_of_month",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_date__step_date_format__day_of_year": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_date__step_date_format__day_of_year",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_numeric_ratios__step_num_ratios_calculate__keep_nominators": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_numeric_ratios__step_num_ratios_calculate__keep_nominators",
                [False, True],
            ),
            "step_preprocess_data__branch_feat_numeric_ratios__step_num_ratios_calculate__keep_denominators": trial.suggest_categorical(
                "step_preprocess_data__branch_feat_numeric_ratios__step_num_ratios_calculate__keep_denominators",
                [False, True],
            ),
        }

        self._pipeline_model.set_params(
            "num_boost_round": 1000,
            "early_stopping_rounds": 200,
            "verbose_eval": 100,
        )
        self._study_preprocess_results = cross_val_score(
            self._pipeline_model,
            self._data_splitted["data"][self.settings.model.name_training]["X"],
            self._data_splitted["data"][self.settings.model.name_training]["y"],
            scoring="roc_auc",
            cv=self._data_splitted["data"][self.settings.model.name_training_cv],
        )

        accuracy = self._study_preprocess_results.mean()
        return accuracy

self._study_preprocess = optuna.create_study(direction="maximize")
self._study_preprocess.optimize(
            self._study_preprocess_search, 
           "n_trials": 15, 
           "timeout": None
        )

```

So I do want to optimize the AUC in my cross-validation process, but I would like to have a chance to select the best trial based on a different condition than `accuracy = self._study_preprocess_results.mean()`. Multi-objective optimization is a cool idea, but I think it would be somehow an overkill in this scenario. 

> optimizing for your main objective (again, loss or other), but saving your secondary metric as a user_attr on the trial. Following the completion of your study, you could then select the top 'X' performers and observer the value of their secondary metric, choosing a more stable configuration. (Using the study => dataframe API should make sorting through trials easier here).

If I understand this correctly return the standard-deviation (or any other custom metric) of the CV scores and then select another trial? I think this would be what I'm looking for.

Diverting from the main topic a bit: I noticed @hvy that optuna offers a generic CV tuner class: https://github.com/optuna/optuna/blob/master/examples/optuna_search_cv_simple.py. Would you recommend using that over the implementation that I presented?

**Comment by blaney83** [comment]
Hey @konradsemsch,

Thanks for the extra info. I _think_ I understand your goal here. For the purpose of this response lets assume your secondary condition is standard deviation (SD) (but as you say it could be anything). Please let me know if I understand this correctly:

You're currently optimizing the mean AUC from your model cross-validation and this is still the objective. However, the best trial is not necessarily the one with the maximum mean AUC, but instead has the lowest e.g. standard deviation and a good mean AUC. Makes sense.

I think you have a few options. @hvy will have to add to this if I miss something, but if you know what your threshold for acceptable

**Comment by blaney83** [comment]
Hey @konradsemsch,

Thanks for the extra info. I _think_ I understand your goal here. For the purpose of this response lets assume your secondary condition is standard deviation (SD) (but as you say it could be anything). Please let me know if I understand this correctly:

You're currently optimizing the mean AUC from your model cross-validation and this is still the objective. However, the best trial is not necessarily the one with the maximum mean AUC, but instead has the lowest SD and a relatively good mean AUC. Makes sense.

I think you have a few options. @hvy will have to add to this if I miss something, but if you know what your threshold for acceptable SD is, you could use a constrained sampler to steer the optimization progress in the correct direction. This would be helpful if you have a model which is prone to over-fitting on a subset of your cv data each time or your study seems to consistently find local maxima with a poor secondary metric.

If you don't have an overfitting/local maxima issue and you just want to optimize for AUC and then select the best trial based on SD, then you could use the user_attr approach you quoted from my original response. For example:

```
        self._study_preprocess_results = cross_val_score(
            self._pipeline_model,
            self._data_splitted["data"][self.settings.model.name_training]["X"],
            self._data_splitted["data"][self.settings.model.name_training]["y"],
            scoring="roc_auc",
            cv=self._data_splitted["data"][self.settings.model.name_training_cv],
        )

        accuracy = self._study_preprocess_results.mean()
        trial.set_user_attr("std_dev", np.std(self._study_preprocess_results))
        return accuracy

```

Following the completion of your trial, you could just filter out trials based on the minimum acceptable AUC (arbitrarily defined by you) and then select the best SD from the subset of trials. You would do this manually by iterating through the trials (`study.trials`) or using the `study.trials_dataframe(...)` method. You could also filter out trials based on a maximum SD you pick and choose the best AUC. It really depends on your domain knowledge about the data pipeline you're optimizing.

Hopefully either of those suggestions works for you. If your studies right now are not containing any trials which have an acceptable SD, then you should look into the Botorch option or even consider MPO. Otherwise, your study likely contains a trial which approaches the desired secondary metric; it's just a matter of tweaking your code to save these intermediate values for post-study analysis.

Cheers!

**Comment by konradsemsch** [comment]
Hey @blaney83! I really appreciate the time you took to explain that - thanks a lot! I think the `user_attr` approach you mentioned is exactly what I'm looking for as it fits into one of the two alternative approaches for selecting the best parameter combination I mentioned in my original question. 

Nevertheless, I will surely also take a look into the other samplers that Optuna offers to explore other approaches 😉
